## Projeye Genel Bakış (Türkçe Özet)

Bu proje, Craigslist New York'tan araç ilanı verilerini çıkarmak için kapsamlı bir web kazıma sürecini göstermektedir. Ana amaç, 'Toyota' araçları hakkında ayrıntılı bilgi toplamak ve bu verileri daha fazla analiz için yapılandırmaktı.

### Neler Yapıldı:
-   Craigslist'ten HTTP istekleri göndermek ve HTML içeriğini almak için Python'ın `requests` kütüphanesi kullanıldı.
-   HTML'i ayrıştırmak ve araç başlıkları, fiyatlar, konumlar gibi belirli veri noktalarını ve bireysel ilan sayfalarından daha ayrıntılı özellikleri (yıl, durum, kilometre, yakıt türü vb.) çıkarmak için `BeautifulSoup` kullanıldı.
-   Çıkarılan verileri yapılandırılmış bir DataFrame'de düzenlemek ve analize hazır hale getirmek için `pandas` kütüphanesi kullanıldı.
-   Sağlam hata işleme (durum kodu kontrolleri) ve nazik kazıma uygulamaları (zaman gecikmeleri) uygulandı.

### Amaç:
-   Dinamik web sayfalarından web kazıma becerilerini ve veri çıkarma tekniklerini sergilemek.
-   Pazar analizi, fiyat tahmini veya envanter bilgileri için temel oluşturabilecek temiz, yapılandırılmış bir kullanılmış Toyota araç ilanı veri seti oluşturmak.
-   Sürecin her adımını ayrıntılı olarak açıklayan, portföy parçası olarak uygun, iyi belgelenmiş ve okunabilir bir Colab not defteri sunmak.

### Neler Oluşturuldu:
-   Craigslist'te gezinebilen, ilgili araç ilanı bilgilerini çıkaran ve birden çok veri noktasını işleyebilen bir Python betiği.
-   'Başlık', 'yıl', 'Fiyat', 'durum', 'yakıt', 'kilometre' ve 'şanzıman' gibi çeşitli özelliklere sahip yaklaşık 300 Toyota araç ilanını içeren bir `pandas` DataFrame.
-   Toplanan verileri harici kullanım için kolayca erişilebilir kılan bir `toyota_crap_data.csv` dosyası.

Bu proje, web kazıma, veri temizleme ve yapılandırma uçtan uca sürecini başarıyla göstermekte ve sonraki veri bilimi görevleri için değerli bir veri seti sağlamaktadır.

## Project Overview (English Summary)

This project demonstrates a comprehensive web scraping process to extract car listing data from Craigslist New York. The primary goal was to collect detailed information about 'Toyota' vehicles and structure this data for further analysis.

### What was done:
-   Utilized Python's `requests` library to send HTTP requests and retrieve HTML content from Craigslist.
-   Employed `BeautifulSoup` to parse the HTML and extract specific data points such as vehicle titles, prices, locations, and more detailed attributes (year, condition, odometer, fuel type, etc.) from individual listing pages.
-   Leveraged the `pandas` library to organize the extracted data into a structured DataFrame, making it ready for analysis.
-   Implemented robust error handling (status code checks) and polite scraping practices (time delays).

### Purpose:
-   To showcase web scraping skills and data extraction techniques from dynamic web pages.
-   To create a clean, structured dataset of used Toyota car listings, which can serve as a foundation for market analysis, price prediction, or inventory insights.
-   To present a well-documented and readable Colab notebook suitable as a portfolio piece, explaining each step of the process in detail.

### What was created:
-   A Python script capable of navigating Craigslist, extracting relevant car listing information, and handling multiple data points.
-   A `pandas` DataFrame containing approximately 300 Toyota car listings with various features like 'Title', 'year', 'Price', 'condition', 'fuel', 'odometer', and 'transmission'.
-   A `toyota_crap_data.csv` file, making the collected data easily accessible for external use.

This project successfully illustrates the end-to-end process of web scraping, data cleaning, and structuring, providing a valuable dataset for subsequent data science tasks.

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

In [2]:
url = "https://newyork.craigslist.org/search/cta?query=toyota"
headers = {"User-Agent": "Mozilla/5.0"}

response = requests.get(url, headers=headers)
print(response.status_code)  # 200 olmalı

soup = BeautifulSoup(response.text, "html.parser")

200


In [5]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import time

# URL'yi temizleyelim
url = "https://newyork.craigslist.org/search/cta?query=toyota"

# Tarayıcı gibi görünmek için daha detaylı bir header ekleyelim
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36",
    "Accept-Language": "tr-TR,tr;q=0.9,en-US;q=0.8,en;q=0.7"
}

response = requests.get(url, headers=headers)

# Eğer sayfa yüklenmediyse hata mesajı basalım
if response.status_code != 200:
    print(f"Hata: Sayfaya erişilemedi. Durum kodu: {response.status_code}")
else:
    soup = BeautifulSoup(response.text, "html.parser")

    # Craigslist yeni yapısında ilanlar genelde 'li' içinde 'cl-static-search-result' class'ına sahip
    # Veya doğrudan başlık ve fiyat etiketlerini arayabiliriz
    cars_elements = soup.find_all("li", class_="cl-static-search-result")

    print(f"İlan sayısı: {len(cars_elements)}")

    cars = []

    for car in cars_elements:
        # Başlık ve Link
        title_tag = car.find("div", class_="title")
        title = title_tag.text if title_tag else "N/A"

        # Fiyat
        price_tag = car.find("div", class_="price")
        price = price_tag.text if price_tag else "N/A"

        # Lokasyon
        location_tag = car.find("div", class_="location")
        location = location_tag.text.strip() if location_tag else "N/A"

        cars.append([title, price, location])

    # DataFrame oluşturma
    df = pd.DataFrame(cars, columns=["Title", "Price", "Location"])
    print(df.head())

    # Veriyi kaydetmek isterseniz:
    # df.to_csv("toyota_ilanlar.csv", index=False)
    display(df)

İlan sayısı: 308
                                               Title    Price  \
0                  2011 Toyota Camry LE 4dr Sedan 6A   $5,600   
1                           Toyota Sienna Wheelchair  $35,000   
2                         Toyota Highlander For Sell   $3,800   
3                           TOYOTA CAMRY LE For Sell   $6,800   
4  2007 Toyota Prius Hybrid Three w/Navigation & ...   $4,250   

             Location  
0  Toyota Camry Sedan  
1            Brooklyn  
2              Queens  
3              Queens  
4            Brooklyn  


Title    Price  \
0                    2011 Toyota Camry LE 4dr Sedan 6A   $5,600   
1                             Toyota Sienna Wheelchair  $35,000   
2                           Toyota Highlander For Sell   $3,800   
3                             TOYOTA CAMRY LE For Sell   $6,800   
4    2007 Toyota Prius Hybrid Three w/Navigation & ...   $4,250   
..                                                 ...      ...   
303  2008 TOYOTA SIENNA LE 8 PASSENGERS FULLY LOADE...   $4,000   
304                  2013 TOYOTA SIENNA LE 8-PASSENGER  $11,500   
305           2013 Toyota Corolla LE Commuter Classic!   $7,927   
306  2021 Toyota Sienna XLE8Passenger XLE 8 Passeng...  $27,195   
307         1992 Toyota 4Runner 🚘 Holidays Special 🌲 🎁   $4,500   

                                    Location  
0                         Toyota Camry Sedan  
1                                   Brooklyn  
2                                     Queens  
3                                     Queens  
4                                   Brooklyn  
..                                       ...  
303                              BROOKLYN NY  
304                            SCARSDALE, NY  
305                                 Harrison  
306  All advertised prices are DOWN PAYMENTS  
307                                 Brooklyn  

[308 rows x 3 columns]

## Kütüphane Açıklamaları

### 1.requests Kütüphanesi

`requests` kütüphanesi, Python'da HTTP istekleri yapmak için kullanılan popüler ve kullanıcı dostu bir kütüphanedir. Bu kütüphane sayesinde web sitelerinden veri çekmek (web scraping), API'lerle etkileşim kurmak veya web servislerine veri göndermek gibi işlemler kolayca gerçekleştirilebilir.

Bu projede `requests` kütüphanesi, Craigslist web sitesinin belirli bir URL'sinden (toyota araması için) HTML içeriğini çekmek amacıyla kullanılmıştır. `requests.get(url, headers=headers)` komutu ile belirtilen URL'ye bir GET isteği gönderilmiş ve web sayfasının HTML içeriği `response.text` olarak alınmıştır. Bu sayede web sayfasının kaynak koduna erişim sağlanmıştır.

### 2.BeautifulSoup Kütüphanesi

`BeautifulSoup` kütüphanesi, HTML ve XML dosyalarından veri çekmek için kullanılan bir Python kütüphanesidir. Web scraping projelerinde, `requests` gibi kütüphanelerle çekilen ham HTML içeriğini ayrıştırmak (parse etmek) ve bu içeriğin içinden belirli etiketlere, class'lara veya id'lere sahip verileri (örneğin başlıklar, fiyatlar, linkler) kolayca bulup çıkarmak için idealdir.

Bu projede `BeautifulSoup` kütüphanesi, `requests` ile elde edilen Craigslist sayfasının HTML içeriğini yapısal olarak anlamak ve ilan bilgilerini (başlık, fiyat, lokasyon) ayıklamak için kullanılmıştır. `BeautifulSoup(response.text, "html.parser")` ile HTML içeriği bir "soup" objesine dönüştürülmüş ve ardından `.find_all()` gibi yöntemlerle istenen veriler hedeflenmiştir.

### 3.pandas Kütüphanesi

`pandas` kütüphanesi, Python'da veri analizi ve manipülasyonu için en güçlü ve yaygın olarak kullanılan kütüphanelerden biridir. Özellikle `DataFrame` adı verilen tablo benzeri veri yapıları sayesinde, farklı kaynaklardan gelen verileri kolayca düzenleyebilir, temizleyebilir, dönüştürebilir ve analiz edebiliriz.

Bu projede `pandas` kütüphanesi, `requests` ve `BeautifulSoup` ile çekilen araba ilan verilerini (başlık, fiyat, lokasyon) düzenli bir `DataFrame` yapısında saklamak ve kullanıcıya okunabilir bir formatta sunmak için kullanılmıştır. Çekilen veriler bir liste içinde toplanmış ve ardından `pd.DataFrame(cars, columns=["Title", "Price", "Location"])` komutu ile bir DataFrame'e dönüştürülmüştür. Bu sayede veriler üzerinde daha kolay inceleme ve ileri analiz yapma imkanı sağlanmıştır.

## Web İstek ve Yanıt İşleme Açıklaması

Bu bölümde, web isteklerinin temel bileşenleri ve bunların nasıl işlendiği açıklanmaktadır.

### 1. URL Tanımı ve Seçimi

**URL (Uniform Resource Locator)**, bir web kaynağının adresidir. İnternet üzerindeki her web sayfasının, resmin veya dosyanın kendine özgü bir URL'si vardır.

`url = "https://newyork.craigslist.org/search/cta?query=toyota"` satırında, web'den veri çekmek istediğimiz Craigslist sayfasının adresi tanımlanmıştır. Bu URL, New York Craigslist üzerinde 'toyota' anahtar kelimesiyle yapılan 'cars & trucks' (`cta`) kategorisindeki bir aramanın sonuçlarını göstermektedir. Bu URL'nin seçilmesinin nedeni, belirli bir kriterdeki ilanlara kolayca erişmek ve bu veriyi kazımak istememizdir.

### 2. Özel HTTP Başlıkları (Headers)

`headers = {"User-Agent": "Mozilla/5.0", ...}` gibi özel HTTP başlıkları, web sunucusuna isteği yapan istemci hakkında bilgi vermek için kullanılır. Özellikle `User-Agent` başlığı, isteği yapan yazılımın (tarayıcı, bot vb.) ne olduğunu belirtir. Birçok web sitesi, varsayılan Python `requests` kütüphanesi User-Agent'ı ile gelen istekleri bot olarak algılayıp engelleyebilir veya kısıtlayabilir. Bu nedenle, bir tarayıcı gibi davranan özel bir `User-Agent` başlığı göndermek, web sitesi tarafından engellenmemek ve normal bir kullanıcı gibi görünmek için kritik öneme sahiptir.

### 3. HTTP İsteği Yapma (`requests.get()`)

`response = requests.get(url, headers=headers)` kodu, belirtilen `url` adresine bir HTTP GET isteği gönderir. Bu istek, tanımladığımız `headers` bilgilerini de içerir. Sunucudan gelen yanıt, `response` objesinde saklanır. Bu obje, sunucunun gönderdiği tüm bilgileri (durum kodu, yanıt içeriği, başlıklar vb.) barındırır.

### 4. Yanıt Durum Kodunu Kontrol Etme (`response.status_code`)

`print(response.status_code)` veya `if response.status_code != 200:` gibi kontroller, HTTP isteğinin başarılı olup olmadığını anlamak için çok önemlidir. Durum kodları, sunucunun isteği nasıl işlediğini gösteren standartlaştırılmış sayılardır:

*   **200 (OK):** İstek başarıyla işlendi ve beklenen yanıt döndürüldü. Bu genellikle işlemin sorunsuz ilerlediği anlamına gelir.
*   **404 (Not Found):** İstenen kaynak sunucuda bulunamadı. URL'nin yanlış yazılmış olması veya kaynağın kaldırılmış olması gibi durumlarda ortaya çıkar.
*   **403 (Forbidden):** Sunucu isteği anladı ancak işleme yetkisi reddedildi. Bu genellikle kimlik doğrulama veya yetkilendirme sorunlarından kaynaklanır ya da sunucunun bot olarak algıladığı isteklere uyguladığı bir engel olabilir.
*   **500 (Internal Server Error):** Sunucuda isteği işlerken beklenmedik bir hata oluştu. Bu, genellikle sunucu tarafındaki bir problemden kaynaklanır.

Durum kodunu kontrol etmek, web kazıma (web scraping) veya API etkileşimlerinde olası hataları erken aşamada tespit etmek ve bu hatalara uygun şekilde tepki vermek için hayati öneme sahiptir. Örneğin, bir 403 hatası alırsanız, `User-Agent` başlığınızı veya IP adresinizi değiştirmeniz gerekebilir.

## HTML Ayrıştırma Mantığı

`requests` kütüphanesi ile web sayfasının HTML içeriği bir metin (`response.text`) olarak alınır. Bu ham HTML metni, doğrudan analiz edilmesi zor, yapılandırılmamış bir dizedir.

`BeautifulSoup(response.text, "html.parser")` satırı, bu ham HTML metnini alıp, `BeautifulSoup` objesine dönüştürür. `"html.parser"` ise HTML'i ayrıştırmak için kullanılan bir ayrıştırıcı belirtir. Bu ayrıştırıcı, HTML dizesini bir ağaç yapısına dönüştürerek, içeriğin hiyerarşik bir şekilde gezilebilir ve aranabilir olmasını sağlar.

Bu yapılandırılmış `soup` objesi sayesinde, web sayfasındaki belirli elementleri kolayca bulabiliriz. Craigslist'teki her bir ilan, genellikle `<li class="cl-static-search-result">` yapısına sahiptir. `soup.find_all("li", class_="cl-static-search-result")` metodu, bu kriterlere uyan tüm `li` etiketlerini bulur ve bunları bir liste olarak `cars_elements` değişkenine atar.

Son olarak, `print(f"İlan sayısı: {len(cars_elements)}")` çıktısı, `cars_elements` listesinin uzunluğunu göstererek, kaç adet ilan elementinin başarıyla bulunduğunu doğrulamamızı sağlar. Bu, web kazıma işleminin beklenen sayıda veri öğesi döndürüp döndürmediğini kontrol etmek için önemlidir.

## Veri Çıkarma Detayları

### Başlık, Fiyat ve Konum Bilgilerinin Ayıklanması

Craigslist sayfasından elde edilen `soup` nesnesi üzerinden, her bir ilan (`car_element`) aşağıdaki adımlarla işlenir:

1.  **İlan Elementleri Üzerinde Döngü:** `soup.find_all("li", class_="cl-static-search-result")` ile bulunan tüm ilan elementleri (`cars_elements`) üzerinde bir `for` döngüsü (`for car in cars_elements:`) başlatılır. Bu döngü, her bir ilanı ayrı ayrı işlememizi sağlar.

2.  **Başlık Bilgisinin Ayıklanması:**
    *   Her bir `car` elementi içinde, ilan başlığı `<div class="title">` etiketinin içinde bulunur.
    *   `title_tag = car.find("div", class_="title")` kodu ile bu `div` etiketi bulunur.
    *   Eğer `title_tag` mevcutsa, `.text` özelliği kullanılarak başlığın metin içeriği alınır: `title = title_tag.text`. Aksi takdirde, `N/A` atanır.

3.  **Fiyat Bilgisinin Ayıklanması:**
    *   Benzer şekilde, ilan fiyatı `<div class="price">` etiketinin içindedir.
    *   `price_tag = car.find("div", class_="price")` kodu ile fiyat `div` etiketi bulunur.
    *   Eğer `price_tag` mevcutsa, `.text` özelliği ile fiyatın metin içeriği alınır: `price = price_tag.text`. Aksi takdirde, `N/A` atanır.

4.  **Konum Bilgisinin Ayıklanması:**
    *   İlanın konumu `<div class="location">` etiketinin içinde yer alır.
    *   `location_tag = car.find("div", class_="location")` kodu ile konum `div` etiketi bulunur.
    *   Eğer `location_tag` mevcutsa, `.text.strip()` özelliği kullanılarak konumun metin içeriği alınır ve baştaki/sondaki boşluklar `strip()` metodu ile temizlenir: `location = location_tag.text.strip()`. Aksi takdirde, `N/A` atanır.

5.  **Bilgilerin `cars` Listesine Eklenmesi:**
    *   Ayıklanan `title`, `price` ve `location` bilgileri bir liste olarak `cars.append([title, price, location])` komutu ile `cars` adlı ana listeye eklenir. Bu sayede, her ilan için elde edilen veriler düzenli bir şekilde saklanır ve daha sonra bir Pandas DataFrame'ine kolayca dönüştürülebilir.

## DataFrame Oluşturma ve Sunumu

Bu bölümde, web'den kazınan verilerin (başlık, fiyat, konum) nasıl bir pandas DataFrame'ine dönüştürüldüğünü ve bu DataFrame'in ne anlama geldiğini açıklıyoruz.

### Verilerin DataFrame'e Dönüştürülmesi

Önceki adımlarda, her bir aracın başlığı, fiyatı ve konumu `cars` adında bir Python listesine eklendi. Bu liste, her bir elemanı `[başlık, fiyat, konum]` şeklinde olan iç listelerden oluşur. Pandas `DataFrame`'i oluşturmak için bu `cars` listesi kullanılır.

```python
df = pd.DataFrame(cars, columns=["Title", "Price", "Location"])
```

*   `pd.DataFrame()`: Bu fonksiyon, pandas kütüphanesinin temel veri yapılarından biri olan DataFrame'i oluşturur. DataFrame'ler, tablolar (Excel veya SQL tabloları gibi) benzeri iki boyutlu, etiketli verilere sahiptir ve çeşitli veri analizi işlemlerini kolaylaştırır.
*   `cars`: Oluşturulacak DataFrame'in verilerini sağlayan listedir.
*   `columns=["Title", "Price", "Location"]`: Bu parametre, DataFrame'deki sütun adlarını belirler. `cars` listesindeki her bir iç listenin sırasıyla "Title", "Price" ve "Location" sütunlarına karşılık gelmesini sağlar. Bu, veriye anlamlı ve erişilebilir bir yapı kazandırır.

### `df` DataFrame'inin Yapısı ve Önemi

Oluşturulan `df` (DataFrame), her bir satırın bir Toyota ilanını temsil ettiği ve her bir sütunun (Title, Price, Location) ilgili özelliklerini içerdiği yapılandırılmış bir veri kümesidir. Bu DataFrame, web kazıma projemizin ana çıktısıdır ve sonraki veri analizi, temizleme, görselleştirme veya makine öğrenimi modelleri için temel teşkil eder.

### `display(df)` Kullanımı

`display(df)` komutu, özellikle Jupyter Notebook veya Google Colab gibi etkileşimli Python ortamlarında bir pandas DataFrame'ini görsel olarak daha düzenli ve kullanıcı dostu bir şekilde görüntülemek için kullanılır. `print(df)`'e kıyasla bazı avantajları vardır:

*   **Gelişmiş Formatlama**: `display()` fonksiyonu, büyük DataFrame'leri sayfalandırır ve kaydırılabilir tablolar halinde sunar, bu da verilerin daha kolay incelenmesini sağlar.
*   **Etkileşimli Çıktı**: Bazı durumlarda, `display()` daha zengin HTML çıktısı sağlayabilir, bu da tabloların sıralanabilir veya aranabilir olmasını mümkün kılabilir (ortamın yeteneklerine bağlı olarak).
*   **Kesme (Truncation)**: Çok fazla satır veya sütun olduğunda, `display()` varsayılan olarak çıktıyı keser (`...` ile gösterir), bu da ekranın taşmasını önlerken önemli bilgileri görmeye olanak tanır. Tam DataFrame'i görmek için ayarlamalar yapılabilir.

Kısacası, `display(df)` web'den topladığımız ve işlediğimiz veriyi daha etkin bir şekilde gözden geçirmemizi ve anlamamızı sağlar.

## Sonuç ve Değerlendirme

Bu projede, Craigslist New York'tan "Toyota" aramasına ait araç ilanları başarıyla çekilmiştir. Toplamda **308 ilan** toplanmış olup, her bir ilan için "Başlık", "Fiyat" ve "Lokasyon" bilgileri elde edilmiştir. Bu veriler, bir Pandas DataFrame'e dönüştürülerek yapılandırılmış bir hale getirilmiştir.

### Potansiyel İçgörüler:

Toplanan bu verilerden çeşitli içgörüler elde edilebilir. Örneğin:
*   **Fiyat Dağılımı:** Toyota modellerinin New York bölgesindeki genel fiyat aralığı ve en uygun/pahalı ilanlar belirlenebilir.
*   **Lokasyon Bazlı Farklılıklar:** Queens, Brooklyn, Staten Island gibi farklı lokasyonlardaki fiyat ortalamaları ve ilan yoğunlukları karşılaştırılarak, hangi bölgelerde daha uygun araçların bulunabileceği veya hangi bölgelerde Toyota ilanlarının daha yaygın olduğu görülebilir.
*   **Popüler Modeller:** Başlıklardan en sık geçen Toyota modelleri (örn. Camry, Prius, Highlander) tespit edilerek hangi modellerin ikinci el piyasasında daha aktif olduğu anlaşılabilir.
*   **Yıl ve Durum Analizi:** Başlık bilgilerinden araçların modelleri ve üretim yılları çıkarılarak, yaşa göre fiyat performans ilişkileri incelenebilir.

### Gelecek Adımlar:

Bu projenin daha da geliştirilmesi için aşağıdaki adımlar düşünülebilir:
*   **Daha Detaylı Veri Çekimi:** İlanların açıklamalarına girilerek daha fazla detay (kilometre, vites tipi, yakıt tipi vb.) çekilebilir.
*   **Veri Temizleme ve Dönüştürme:** Fiyat sütunundaki '$' ve ',' gibi karakterlerin kaldırılması ve sayısal tipe dönüştürülmesi, lokasyon verilerinin standartlaştırılması gibi işlemler yapılabilir.
*   **Görselleştirmeler:** Toplanan veriler kullanılarak fiyat dağılımı, lokasyona göre fiyat farklılıkları, en popüler modellerin çubuk grafikleri gibi interaktif görselleştirmeler oluşturulabilir.
*   **Farklı Sitelerden Veri Entegrasyonu:** Craigslist dışındaki diğer ikinci el araç sitelerinden (örn. Facebook Marketplace, Autotrader) benzer veriler çekilerek daha kapsamlı bir veri seti oluşturulabilir ve pazarın genel görünümü daha iyi analiz edilebilir.
*   **Makine Öğrenmesi Uygulamaları:** Yeterli ve temiz veri toplandıktan sonra, araç fiyatlarını tahmin eden veya en iyi fırsatları belirleyen bir makine öğrenmesi modeli geliştirilebilir.

### Daha geniş kapsamlı bir veri çekme işlemi yapalım.

In [10]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np
import time

# --------------------------------------------------
# AYARLAR
# --------------------------------------------------
url = "https://newyork.craigslist.org/search/cta?query=toyota"
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
}

# --------------------------------------------------
# İLAN LİNKLERİNİ TOPLA
# --------------------------------------------------
print("Linkler toplanıyor...")
response = requests.get(url, headers=headers)
soup = BeautifulSoup(response.text, "html.parser")

links = [
    a["href"]
    for a in soup.select("li.cl-static-search-result a")
    if a.get("href")
]

print(f"{len(links)} adet ilan bulundu.")

# --------------------------------------------------
# İLAN DETAYLARINI ÇEK
# --------------------------------------------------
all_cars = []

for i, link in enumerate(links[:100]):  # ilk 100 ilan
    try:
        print(f"{i + 1}. ilan detayları okunuyor...")
        res = requests.get(link, headers=headers)
        s = BeautifulSoup(res.text, "html.parser")

        car_info = {"Link": link}

        # Başlık
        title_tag = s.find("span", id="titletextonly")
        car_info["Title"] = title_tag.text if title_tag else np.nan

        # Fiyat
        price_tag = s.find("span", class_="price")
        car_info["Price"] = (
            price_tag.text.replace("$", "").replace(",", "")
            if price_tag else np.nan
        )

        # Özellikler (attrgroup)
        for attr in s.select(".attrgroup .attr"):
            label = attr.find(class_="label")
            value = attr.find(class_="value")

            if label and value:
                key = label.text.replace(":", "").strip().lower()
                val = value.text.strip()
                car_info[key] = val
            elif ":" in attr.text:
                parts = attr.text.split(":", 1)
                car_info[parts[0].strip().lower()] = parts[1].strip()

        all_cars.append(car_info)
        time.sleep(1)

    except Exception as e:
        print("Hata:", e)
        continue

# --------------------------------------------------
# DATAFRAME OLUŞTUR
# --------------------------------------------------
df = pd.DataFrame(all_cars)

# --------------------------------------------------
# TEMİZLİK
# --------------------------------------------------
# Boşluk / None → NaN
df = df.replace(r'^\s*$', np.nan, regex=True)

# 🔹 TITLE'DAN ARAÇ YILI ÇIKAR
df["year"] = df["Title"].str.extract(r'(19\d{2}|20\d{2})')
df["year"] = pd.to_numeric(df["year"], errors="coerce")

# --------------------------------------------------
# SADECE GEREKLİ SÜTUNLAR
# --------------------------------------------------
istenenler = [
    "Title",
    "year",
    "Price",
    "condition",
    "cylinders",
    "drive",
    "fuel",
    "odometer",
    "paint color",
    "transmission",
    "type"
]

df_final = df.reindex(columns=istenenler)

# --------------------------------------------------
# SAYISAL DÖNÜŞÜMLER
# --------------------------------------------------
df_final["Price"] = pd.to_numeric(df_final["Price"], errors="coerce")

df_final["odometer"] = (
    df_final["odometer"]
    .str.replace(",", "", regex=False)
    .astype(float)
)

# Mantıksız yılları temizle
df_final.loc[
    (df_final["year"] < 1980) | (df_final["year"] > 2025),
    "year"
] = np.nan

# --------------------------------------------------
# SONUÇ
# --------------------------------------------------
print("\nTablo Güncellendi!")
display(df_final.head(10))

# --------------------------------------------------
# CSV KAYDET
# --------------------------------------------------
df_final.to_csv(
    "toyota_crap_data.csv",
    index=False,
    encoding="utf-8-sig"
)

print("Veri seti 'toyota_crap_data.csv' olarak kaydedildi!")


Linkler toplanıyor...
307 adet ilan bulundu.
1. ilan detayları okunuyor...
2. ilan detayları okunuyor...
3. ilan detayları okunuyor...
4. ilan detayları okunuyor...
5. ilan detayları okunuyor...
6. ilan detayları okunuyor...
7. ilan detayları okunuyor...
8. ilan detayları okunuyor...
9. ilan detayları okunuyor...
10. ilan detayları okunuyor...
11. ilan detayları okunuyor...
12. ilan detayları okunuyor...
13. ilan detayları okunuyor...
14. ilan detayları okunuyor...
15. ilan detayları okunuyor...
16. ilan detayları okunuyor...
17. ilan detayları okunuyor...
18. ilan detayları okunuyor...
19. ilan detayları okunuyor...
20. ilan detayları okunuyor...
21. ilan detayları okunuyor...
22. ilan detayları okunuyor...
23. ilan detayları okunuyor...
24. ilan detayları okunuyor...
25. ilan detayları okunuyor...
26. ilan detayları okunuyor...
27. ilan detayları okunuyor...
28. ilan detayları okunuyor...
29. ilan detayları okunuyor...
30. ilan detayları okunuyor...
31. ilan detayları okunuyor...
32.

Title    year    Price  \
0                                  2010 TOYOTA CAMRY  2010.0   3200.0   
1                                  2010 TOYOTA CAMRY  2010.0    999.0   
2  2021 TOYOTA COROLLA LE PRESTINE COND 65K MI 1 ...  2021.0      1.0   
3                  2011 Toyota Camry LE 4dr Sedan 6A  2011.0   5600.0   
4                           Toyota Sienna Wheelchair     NaN  35000.0   
5                         Toyota Highlander For Sell     NaN   3800.0   
6                           TOYOTA CAMRY LE For Sell     NaN   6800.0   
7  2007 Toyota Prius Hybrid Three w/Navigation & ...  2007.0   4250.0   
8    2010 Toyota Prius Hybrid. Super Low Mileage 40K  2010.0   9990.0   
9                                2012 Toyota Corolla  2012.0   1800.0   

   condition    cylinders drive    fuel  odometer paint color transmission  \
0        NaN          NaN   NaN     gas  100000.0      silver    automatic   
1        NaN          NaN   NaN     gas   80000.0       black    automatic   
2   like new  4 cylinders   fwd     gas   65000.0         NaN    automatic   
3        NaN  4 cylinders   fwd     gas  145425.0       white    automatic   
4  excellent  6 cylinders   fwd     gas   69500.0         NaN    automatic   
5  excellent  6 cylinders   4wd     gas  128000.0      silver    automatic   
6   like new  4 cylinders   fwd     gas  104000.0        blue    automatic   
7  excellent  4 cylinders   fwd     gas  197800.0         NaN    automatic   
8  excellent          NaN   NaN  hybrid   40496.0         NaN    automatic   
9   like new  4 cylinders   fwd     gas  125000.0       black    automatic   

        type  
0      sedan  
1      sedan  
2      sedan  
3      sedan  
4    minivan  
5        SUV  
6      sedan  
7  hatchback  
8  hatchback  
9      sedan

Veri seti 'toyota_crap_data.csv' olarak kaydedildi!


In [14]:
from google.colab import files
files.download('toyota_crap_data.csv') #indirmek için de bu alanı çalıştır

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Genişletilmiş Veri Çekme ve Detaylı Analiz

Önceki adımlarda sadece ana arama sayfasından başlık, fiyat ve lokasyon gibi temel bilgiler çekilmişti. Ancak web kazıma projelerinde, her bir ilanın kendi sayfasında yer alan çok daha detaylı bilgilere ulaşmak genellikle gereklidir. Bu bölümde, her bir ilan linkine tek tek girerek (ilk 100 ilan için), aracın durumu, silindir sayısı, yakıt tipi, kilometre, renk, şanzıman ve araç tipi gibi ek özellikleri çekme ve bu veriyi yapılandırma süreci açıklanmaktadır.

### 1. İlan Linklerinin Toplanması

İlk olarak, arama sonuçları sayfasındaki her bir ilan için ayrıntılı sayfanın linkleri toplanır. `soup.select("li.cl-static-search-result a")` ile ilan kartlarının içindeki bağlantılar (a etiketleri) bulunur ve `href` nitelikleri `links` listesine eklenir. Bu sayede her bir ilanın benzersiz detay sayfasına ulaşmak için bir köprü oluşturulur.

### 2. İlan Detaylarının Çekilmesi

Toplanan `links` listesi üzerinde bir döngü başlatılır. Her bir link için:

*   **Yeni HTTP İsteği**: `requests.get(link, headers=headers)` ile ilanın detay sayfasına yeni bir HTTP GET isteği gönderilir. Bu adım, her ilanın kendine ait zengin içerikli HTML'ine erişmemizi sağlar.
*   **Detay Sayfası Ayrıştırma**: Elde edilen HTML içeriği, yine `BeautifulSoup` kullanılarak ayrıştırılır ve bir `BeautifulSoup` objesi (`s`) oluşturulur.
*   **Bilgi Ayıklama**: İlan başlığı (`<span id="titletextonly">`), fiyat (`<span class="price">`) ve `attrgroup` sınıfına sahip özellik gruplarından (örneğin, "condition", "cylinders", "odometer" gibi) ek detaylar ayıklanır. Bu özellikler genellikle etiket ve değer çiftleri olarak düzenlenmiştir.
*   **Veri Yapılandırma**: Ayıklanan tüm bilgiler `car_info` adında bir sözlükte depolanır ve ardından bu sözlük `all_cars` listesine eklenir. `time.sleep(1)` komutu, sunucuya art arda çok fazla istek göndermemek ve IP adresimizin engellenmesini önlemek için her istek arasında 1 saniyelik bir bekleme süresi ekler.

### 3. DataFrame Oluşturma ve Temizleme

`all_cars` listesi, her biri bir ilana ait detayları içeren sözlüklerden oluşur. Bu liste, `pd.DataFrame(all_cars)` kullanılarak bir pandas DataFrame'ine dönüştürülür.

Ek olarak, veriler üzerinde bazı temizleme ve dönüştürme işlemleri yapılır:

*   **Boş Değerlerin Temizlenmesi**: DataFrame'deki boşluk içeren veya 'None' olan hücreler `np.nan` (Not a Number) ile değiştirilir. Bu, eksik veri analizi ve sayısal işlemler için önemlidir.
*   **Yıl Bilgisinin Çıkarılması**: İlan başlıklarından düzenli ifadeler (`r'(19\d{2}|20\d{2})'`) kullanılarak araçların üretim yılları (`year`) çıkarılır ve sayısal tipe dönüştürülür.
*   **Sayısal Dönüşümler**: 'Price' ve 'odometer' (kilometre) sütunlarındaki metinsel değerler (örn. '$', ',' gibi karakterler temizlenerek) sayısal tipe dönüştürülür. Geçersiz veya anlamsız yıl değerleri (`<1980` veya `>2025`) `np.nan` olarak ayarlanır.

### 4. Sonuç DataFrame'i (`df_final`)

Son olarak, analiz için en önemli sütunları içeren `df_final` adında yeni bir DataFrame oluşturulur. Bu DataFrame, her bir ilanın daha zengin ve temizlenmiş özellik setini sunar ve aşağıdaki sütunları içerir:

*   **Title**: İlan başlığı.
*   **year**: Araç üretim yılı (başlıktan çıkarıldı).
*   **Price**: Aracın fiyatı (sayısal).
*   **condition**: Aracın durumu (örn. 'good', 'excellent').
*   **cylinders**: Silindir sayısı.
*   **drive**: Çekiş tipi (örn. 'fwd', 'rwd').
*   **fuel**: Yakıt tipi (örn. 'gas', 'electric').
*   **odometer**: Kilometre bilgisi (sayısal).
*   **paint color**: Boya rengi.
*   **transmission**: Şanzıman tipi (örn. 'automatic', 'manual').
*   **type**: Araç tipi (örn. 'sedan', 'suv').

Bu `df_final` DataFrame, çok daha detaylı ve analiz edilebilir bir veri seti sağlayarak, New York Craigslist'teki Toyota ilanları hakkında derinlemesine içgörüler elde etme imkanı sunar.